[Gstreamer][WebRTC] webrtc/video-disabled-black.html fails

philn opened the pull request:
#### d9e6730c373fca89a0a9a11c457125805c749470
<pre>
[Gstreamer][WebRTC] webrtc/video-disabled-black.html fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=254212">https://bugs.webkit.org/show_bug.cgi?id=254212</a>
&lt;rdar://problem/107046108&gt;

Reviewed by Xabier Rodriguez-Calvar.

When disabling an incoming WebRTC track the mediastreamsrc now updates and propagates the
corresponding GstStream. This allows decodebin3 to properly reconfigure itself, by getting rid of
the decode chain (the disabled stream is not encoded).

This patch also disables the `parse-streams` property on urisourcebin for MSE and MediaStreams, it&apos;s
needed mostly for buffering adaptivedemux-based streams anyway. Parsing will then be done
downstream, by decodebin3, which also simplifies input stream reconfigurations that are now managed
internally by decodebin3.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::configureElement):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(webkitMediaStreamNew):
(webkitMediaStreamSrcPostStreamCollection):

Canonical link: <a href="https://commits.webkit.org/266589@main">https://commits.webkit.org/266589@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21b59d4f64dbc6e5929ea317801576c097815f65

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14210 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14522 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14860 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15945 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13461 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14603 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16131 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14951 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16665 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12235 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12810 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19827 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13313 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12974 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16179 "1 api test failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13525 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11381 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12804 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3440 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17140 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13369 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -157,6 +163,23 @@ class InternalSource final : public MediaStreamTrackPrivate::Observer,
+        gst_pad_add_probe(pad.get(), GST_PAD_PROBE_TYPE_QUERY_UPSTREAM, reinterpret_cast<GstPadProbeCallback>(+[](GstPad*, GstPadProbeInfo* info, InternalSource*) -> GstPadProbeReturn {
+            auto* query = GST_PAD_PROBE_INFO_QUERY(info);
+            switch (GST_QUERY_TYPE(query)) {
+#if GST_CHECK_VERSION(1, 22, 0)
```
Considering that we are only reacting on QUERY_SELECTABLE and no-op otherwise, wouldn't it to make more sense flag they whole probe?

philn committed:
[Gstreamer][WebRTC] webrtc/video-disabled-black.html fails
https://bugs.webkit.org/show_bug.cgi?id=254212
<rdar://problem/107046108>

Reviewed by Xabier Rodriguez-Calvar.

When disabling an incoming WebRTC track the mediastreamsrc now updates and propagates the
corresponding GstStream. This allows decodebin3 to properly reconfigure itself, by getting rid of
the decode chain (the disabled stream is not encoded).

This patch also disables the `parse-streams` property on urisourcebin for MSE and MediaStreams, it's
needed mostly for buffering adaptivedemux-based streams anyway. Parsing will then be done
downstream, by decodebin3, which also simplifies input stream reconfigurations that are now managed
internally by decodebin3.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::configureElement):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
(webkitMediaStreamNew):
(webkitMediaStreamSrcPostStreamCollection):

Canonical link: https://commits.webkit.org/266589@main
